### ISSUE TITLE
improve documentation  system (for contributors to web site)

### ISSUE THREAD

**Issue by ghost** [DlangScience/DlangScience.github.io#7]
Since we don't yet have a mailing list for DlangScience, perhaps we can temporarily use issues under DlangScience.github.io as a kind of discussion forum. This issue could be dedicated to further improvements to the web-site setup or even simple clarifications of any problems encountered.

**Comment by ghost** [comment]
Copied from Gitter: "I had problems installing/upgrading Jekyll on Ubuntu 14.04 because the ubuntu jekyll package is too old and the default ruby is still 1.9.3. Basically had to use gem2.0 install and not use bundler to get things working. Also had to install a runtime for execjs which was not installed by default . I also had to tweak some dub dependencies because ddox didn't play well with the latest dmd/vibe-d combo. But now it's working, whew!"

**Comment by John-Colvin** [comment]
Leaving aside the Jekyll stuff for now (I have no idea about it really, I had assumed that seeing as it is common software and endorsed by github it would work out of the box  on most systems), what dub dependencies needed changing and what sort of errors were you getting before you changed them?

**Comment by ghost** [comment]
What is the correct process for adding new documentation sub-pages in contents/main? I have done this:
- updated my clone of DlangScience.github.io and created a feature branch
- fired up `jekyll serve`
- edited content/main/index.md and added link `[contributing](contrib_help.html)` 
- created new file content/main/contrib_help.md (with header)

```
  ---
  layout: default
  ---
```
- ran `./get_pages.sh`

Next steps. Push to my fork in github and create a pull request.
Is this okay?

**Comment by ghost** [comment]
@John-Colvin building ddox automatically used vibe-d-0.7.23 which in turn used a version of libevent that I don't remember but which had compile errors with dmd 2.068.0. I tweaked the dub.json of ddox to require vibe-d-0.7.24, removed the packages that had been automatically fetched and rebuilt ddox.

**Comment by John-Colvin** [comment]
`get_pages.sh` will have generated a new html file `contrib_help.html`, which will need to be commited, along with `content/main/contrib_help.md`. `git status` should tell you what is staged, unstaged or untracked. They will probably be untracked at the moment, so run `git add contrib_help.html content/main/contrib_help.md` to stage them, then commit the changes, push to your fork and make a pull request.

FYI

```
---
layout: default
---
```

is added automatically by `get_pages.sh` if it doesn't find a header.

**Comment by John-Colvin** [comment]
> building ddox automatically used vibe-d-0.7.23 which in turn used a version of libevent that I don't remember but which had compile errors with dmd 2.068.0. I tweaked the dub.json of ddox to require vibe-d-0.7.24, removed the packages that had been automatically fetched and rebuilt ddox.

Did you try just clearing out everything in `~/.dub/packages`? I don't fully understand how/when dub decides to get more recent versions, but giving a clean slate seems to help me get things working sometimes.

**Comment by ghost** [comment]
The `---\nlayout: default\n---` added automatically by get_pages.sh still didn't work correctly. I got a page of simple html with that text at the top but no correct html header.

**Comment by ghost** [comment]
> > building ddox automatically used vibe-d-0.7.23 which in turn used a version of libevent that I don't remember but which had compile errors with dmd 2.068.0. I tweaked the dub.json of ddox to require vibe-d-0.7.24, removed the packages that had been automatically fetched and rebuilt ddox.
> 
> Did you try just clearing out everything in `~/.dub/packages`? I don't fully understand how/when dub decides to get more recent versions, but giving a clean slate seems to help me get things working sometimes.
> 
> Yes, I tried that, but ddox insisted on vibe-d-0.7.23.
> Presumably this is a temporary dependency problem given that the new dmd/vibe-d combo are just barely released.

**Comment by John-Colvin** [comment]
> The ---\nlayout: default\n--- added automatically by get_pages.sh still didn't work correctly. I got a page of simple html with that text at the top but no correct html header.

That's very odd. The end result (i.e. the html that Jekyll sees) should be an html file with

```
---
layout: default
---
```

as the first 3 lines. The automatically added header is used for all the individual repository pages. Do they work OK for you locally?

**Comment by ghost** [comment]
No, the individual repository pages look like this attachment:
![screenshot_2015-08-11_11-59-00](https://cloud.githubusercontent.com/assets/13017659/9195103/9fc2a79a-4020-11e5-9ce9-349e99aee3e3.png)

**Comment by John-Colvin** [comment]
> Yes, I tried that, but ddox insisted on vibe-d-0.7.23.
> Presumably this is a temporary dependency problem given that the new dmd/vibe-d combo are just barely released.

https://github.com/rejectedsoftware/ddox/issues/90

**Comment by John-Colvin** [comment]
https://github.com/DlangScience/DlangScience.github.io/pull/8 should fix the header problem

**Comment by John-Colvin** [comment]
I've had a bit more of a thought about this:

If it's a pain to deal with all the different files that get altered etc, I'm fine with pull requests of only non-auto-generated files, e.g. a new/altered markdown file, a change in _layout/default.html.

**Comment by ghost** [comment]
That would be great. Shouldn't all the (locally?) auto-generated files be in directories listed in .gitignore so they don't create so many merge conflicts?

**Comment by John-Colvin** [comment]
I need to be able to commit the auto-generated files, so .gitignore won't work. Just use `git add <whateverFilesYouWant>` to manually add the relevant ones and leave all the auto-generated stuff unstaged. Then `git commit -m'<myCommitMessage>'` (note no `-a` there).

**Comment by John-Colvin** [comment]
What you've got at https://github.com/jansengb/DlangScience.github.io/commit/47dc3cec9ddd50cf533fbce2dddcafe74a07a611 looks good to me, please make a pull request

**Comment by jansengb** [comment]
Okay, just want remove line breaks in paragraphs first.

**Comment by ghost** [comment]
@John-Colvin I've updated my local clone and tried out some of your new-doc-gen changes. Peeked at dstat.summary doc - looks fixed so great! 

I still struggle with removing files added locally by get_pages.sh before I can merge your latest updates. I'm using `git stash -a`  and `git stash clear` which seems to work okay. `git status` still shows untracked content in the content/repo/package directories (.dub, __dummy.html, docs.json and dub.selections.json), which don't conflict with the merge. Maybe these could be removed automatically at end of the loop in get_pages.sh.

**Comment by John-Colvin** [comment]
dstats is actually live with the new system, I went ahead and merged it myself because it was actually broken: http://dlangscience.github.io/dstats/api/dstats/summary/geometricMean.html

Assuming I've saved anything I want to keep (e.g. it's committed or stashed), I often just do `rm -rf *` to delete everything except the hidden files in the root of the package (so it doesn't touch `.git`), then `git reset --hard` to get the checked-in content back, then `git pull` to get up-to-date. In the DlangScience.github.io case you'd want to do `git submodule update --init && git submodule update --remote` as well.
